src/industry_gui.cpp

```diff
@@ -1430,12 +1440,26 @@ class IndustryDirectoryWindow : public Window {
 	 */
 	static int GetCargoTransportedSortValue(const Industry *i)
 	{
-		int p1 = GetCargoTransportedPercentsIfValid(i, 0);
-		int p2 = GetCargoTransportedPercentsIfValid(i, 1);
-
-		if (p1 > p2) Swap(p1, p2); // lower value has higher priority
+		CargoID filter = IndustryDirectoryWindow::produced_cargo_filter;
+
+		int percentage = 0, produced_cargo_count = 0;
+		for (uint id = 0; id < lengthof(i->produced_cargo); id++) {
+			if (filter == CF_NONE) {
+				return percentage;
+			} else if (filter == CF_ANY) {
+				int transported = GetCargoTransportedPercentsIfValid(i, id);
+				if (transported != -1) produced_cargo_count++;
+				percentage += transported == -1 ? 0 : transported;
+				if (produced_cargo_count == 0 && id == lengthof(i->produced_cargo) - 1 && percentage == 0) {
+					return transported;
+				}
+			} else if (filter == i->produced_cargo[id]) {
+				return GetCargoTransportedPercentsIfValid(i, id);
+			}
+		}
```

Comment:
This for loop (and the one in the sorter) is weird - it still enters the loop even if `filter == CF_NONE` which is unconnected to the loop

Reply:
`filter == CF_NONE` is now outside the two `for `loops.